src/java/org/apache/cassandra/service/accord/AccordStateCache.java

```diff
@@ -149,17 +145,12 @@ public NamedMap(String name)
         }
     }
 
-    public final Map<Object, Node<?, ?>> active = new HashMap<>();
     private final Map<Object, Node<?, ?>> cache = new HashMap<>();
-    private final Map<Object, WriteOnlyGroup<?, ?>> pendingWriteOnly = new HashMap<>();
-    private final Set<Instance<?, ?>> instances = new HashSet<>();
-
-    private final NamedMap<Object, Future<?>> loadFutures = new NamedMap<>("loadFutures");
-    private final NamedMap<Object, Future<?>> saveFutures = new NamedMap<>("saveFutures");
+    private final Set<Instance<?, ?, ?>> instances = new HashSet<>();
 
-    private final NamedMap<Object, Future<Data>> readFutures = new NamedMap<>("readFutures");
-    private final NamedMap<Object, Future<?>> writeFutures = new NamedMap<>("writeFutures");
+    private final NamedMap<Object, AsyncResult<Void>> saveResults = new NamedMap<>("saveResults");
 
+    private int linked = 0;
```

Comment:
that's misleading, they're not neccesarily eligibleForEviction. `unreferenced` is probably clearer than linked though

Reply:
Doesn’t unreferenced basically mean eligible for eviction (if last in queue)? But yeah, unreferenced is an improvement

Reply:
Almost. To be eligible for eviction, there also needs to be no unfinished operations, it needs to be in the right state, etc.

Reply:
Should we even be inserting into the list if we aren’t yet eligible for collection? Doesn’t that cause inefficiencies when actually evicting?

Reply:
It would, though I’d expect it to be negligible in practice, and be both less impactful and preferable to the added complexity of a interim cache tier.

Reply:
What’s your thought process that expects this to be negligible? I think we just need a variation in flush performance, and the slowest will become the oldest and have to be jumped over on every eviction. In the worst case the cache could be both full and empty, and involve lots of pointer chasing to do nothing.

That said, I’m not sure how this interplays with planned changes to persistence.

Does there need to be an intermediate tier though? Either we would be in neither until eligible, or still would be to just have an extra link in the direction used for eviction, that is only set once the item is actually eligible (that way we correctly track hotness).

Reply:
So from the expected cache behavior perspective, we should be inserting into this list as soon as it becomes unreferenced, because it’s tracking the least recently used objects

Jumping over a handful of stragglers isn’t optimal, but still qualifies as negligible I think. 
The full/empty case is more of a concern though, and something that would be more likely with smaller cache sizes. I haven't thought of any solutions that don't cause their own problems. 
The extra link idea doesn’t really work because the cache passively updates it’s state by iterating over the linked list, so that additional link would never be made.

WDYT about calling maybeEvict periodically, instead as part of reference/release? That would at least bound the amount of wasted work in extreme cases without adding any complexity to the cache

Reply:
Wouldn't that lead to fairly inefficient use of the available memory?

If we're looking this up lazily, how are we clearing the `saveResults` map? A quick glance suggests this is also done lazily? Are we accounting for the memory we're consuming in this map?

It doesn't look like we're clearing memory associated with the mutations when they complete, and it also looks like we can accumulate an unbounded number of them against a single key, so that we perhaps can have situations where we _never_ reclaim memory?

I'll dig into this a bit more tomorrow, but I feel like we may want to be a bit more precise with this.

Reply:
> Wouldn't that lead to fairly inefficient use of the available memory?

You mean calling maybeEvict every minute?

> If we're looking this up lazily, how are we clearing the saveResults map?

In the common case, it happens once the safe future completes

> Are we accounting for the memory we're consuming in this map?

No

Reply:
> In the common case, it happens once the safe future completes

Ah, I see where it happens now. I don't think it can be sure to clear it though, as we merge later results with earlier ones into a single result. If we keep merging new save results before the prior one completes, do we accumulate memory indefinitely?

But as far as maintaining the broader eviction strategy is concerned, wouldn't this provide the perfect hook to update an additional pointer that permits us to maintain the map efficiently?

Reply:
> do we accumulate memory indefinitely?

We might be OK here actually, chains of `AsyncChainCombiner` may naturally flatten over time. Will take another look in the morning. Spotted a potential issue with `AsyncChainCombiner` though, will mention separately.

Reply:
ATM `canEvict` is defined as:

1) node.references == 0 (also known as linked)
2) node.isLoaded // i feel this isn't needed, this happens only in the case of failed operations.  if the load also failed we want to evict right away and we already do that in `org.apache.cassandra.service.accord.AccordStateCache.Instance#release`, so these should only be Pending
3) no pending saveResults

So, lets say we remove #2, then the only case is #3 which is unsafe to evict until we know the result (failed saves can not be evicted...)...  

So, if we don't `push` and instead do something as follows

```
public void release(S safeRef)
...
            maybeEvict();
        }

        private void maybePush(Node<K, V> node)
        {
            if (node.references != 0) return;
            Invariants.checkState(node.isUnlinked());
            AsyncResult<Void> save = getAsyncResult(saveResults, node.key());
            if (save == null)
            {
                logger.trace("Moving {} from active pool to cache", node.key());
                push(node);
            }
            else if (!save.isDone())
            {
                save.addCallback((s, f) -> {
                    if (f != null)
                    {
                        //TODO what do we do here?  Do we just leave in cache?  That is kinda what happens anyways
                        return;
                    }
                    maybePush(node);
                }, TODONeedRefToExecutor());
            }
            else if (!save.isSuccess())
            {
                //TODO what do we do here?  Do we just leave in cache?  That is kinda what happens anyways
            }
            else
            {
                // save was success, but was another one added?  Was the node refereneced?
                maybePush(node);
            }
        }
```

This adds more complexity to things, but keeps non-evicitable nodes out of the list... Now, if we are full... we are full and most likely have other issues.  This logic doesn't prevent us from getting full as we just ignore nodes that failed to write, building up until we OOM (at least OOM from a cache point of view)...

> That said, I’m not sure how this interplays with planned changes to persistence.

I do not know what is planned, but right now our metadata is only durable if `saveResults` is empty or was successful, so if we have another system for durability, then we don't care about #3 either I believe... so all we care about is reference counts?  So how much we care about this case I guess is a function of what this planned work does in the face of failures?

Reply:
> But as far as maintaining the broader eviction strategy is concerned, wouldn't this provide the perfect hook to update an additional pointer that permits us to maintain the map efficiently?

Could you elaborate on this?

Reply:
Something like this? (with the AsyncChain issue addressed too): https://github.com/belliottsmith/cassandra/tree/immutable-state-suggest

DISCLAIMER: untested and not thought about super carefully, just showing the general shape of what I think would work for a fairly simple change that would address this issue without much changing the efficiency otherwise.

Reply:
I think I get the gist of what you’re proposing here. A few things I noticed:

First, we can’t start evicting from the tail, since the tail may not be evictable. So we’d need to either maintain a separate tail, or start scanning from the tail to find the first evictable node. Scanning from the tail has the same efficiency problem, at least until we find the first evictable node. Maintaining a separate tail would be possible, but we’d have to store some additional state on the node to indicate which was pushed onto the queue first, so older nodes that become evictable would know they should replace the current eviction tail.

First, this is less resilient against failures and bugs. The lazy evaluation approach currently used only requires that a node be unreferenced, and will evict on it’s own as the save results complete, without additional input from AsyncOperation, which simplifies failures handling and serves as a built in escape hatch from some leaking bugs. Without some additional maintenance, evictable nodes may never make their way into the eviction queue and be kept in memory forever.

Second, since this relies on the AsyncOperation completing to mark nodes as evictable, nothing from a large txn becomes evictable until all write operations have completed. With lazy evaluation, they’re evictable as soon as their writes complete.

Also disclaimer: I’m reading and responding in the time between 2 appointments, so may have missed something important.

Reply:
I think we could save the lazy evaluation by keeping a pendingSaveResult -> Node map and updating as appropriate when maybeEvict is called.

Reply:
...and if we used a TreeSet or something to store evictable nodes sorted by when they were "released", we could evict by pulling objects off the tail instead of maintaining a parallel linked list

Reply:
I think the tail issue is resolved if we have a dummy head pointer as proposed elsewhere that is both the head and tail?

Regarding bugs, we could periodically look at the full queue and report if there’s an element we should have found via the other queue? 

the nice thing about the extra linked-list is that it’s very cheap to maintain, I’d prefer not to keep another map (we already have quite a few). But let me ponder your point about delays to eviction from the combined save process, that is suboptimal.

Reply:
> I think the tail issue is resolved if we have a dummy head pointer as proposed elsewhere that is both the head and tail?

I don't see how? The dummy head seems to be more stylistic than functional, and doesn't seem to address either of the issues raised with starting from the tail.

Reply:
dummy heads that are also the tail slightly simplify doubly-linked lists but also since they’re dummies the tail would always precede the first actual evictable entry, as it would not itself be an entry

Reply:
I see. So the issue was not finding the tail of the full list, but that the tail may not be evictable, and the issues caused by that:

> First, we can’t start evicting from the tail, since the tail may not be evictable. So we’d need to either maintain a separate tail, or start scanning from the tail to find the first evictable node. Scanning from the tail has the same efficiency problem, at least until we find the first evictable node. Maintaining a separate tail would be possible, but we’d have to store some additional state on the node to indicate which was pushed onto the queue first, so older nodes that become evictable would know they should replace the current eviction tail.

Reply:
If the dummy head were itself also the tail, then it would also have its own prevEvictable field that would be correctly maintained as normal

Reply:
Ok so you’re saying that this maintains it’s own tail then, which I could see happening with the dummy head. It doesn’t fix the problem though. As written, when cleanup sets prevEvictable, it sets the prevEvictable pointer on `next`, but if `prevEvictable` isn’t set on `next`, it’s still stranded from an eviction perspective until a neighbor becomes evictable. This could be solved by making the evictable list a doubly linked list.... But we’ll still have to do some pointer chasing to find the closest evictable node when a new node becomes evictable

Reply:
Thinking about this some more, I think this is getting too overwrought for what the state cache does and the problem we’re talking about. I think we should just call maybeEvict once a minute and move on, I think that’s roughly good enough, and will leave the state cache in a place where it will be much easier to fix any perf issues we encounter during testing.

Reply:
Once a minute is definitely not enough. You do not need much traffic to exhaust the cache in that time interval. Let's say you have 128MiB cache, 5k qps, you only 450 bytes in cache per query to exhaust the cache. Just 50 bytes per cache wastes 10% of your cache space.

I'm strongly -1 on timed cache clearing. Cache management should dynamic based on actual demand and conditions. If we decide to accept the pointer chasing cost that's one thing, but a cache clearance frequency is another tunable we _shouldn't_ have, and we have enough of those.

I don't mind punting on this, with a `TODO (expected)` to revisit, perhaps when we address persistence. But not with a periodic clearance, let's keep doing this on demand.

Reply:
There remain some other very simple possibilities, such as dequeuing the tail elements that we want to evict, and place them in a separate queue for eviction as soon as they are ready, so that we don't repeatedly walk over them.

Reply:
Ok one minute was a bit of an exaggeration. I am in favor of accepting the pointer chasing cost until we can demonstrate it's a bottleneck though.

Reply:
I'm happy to punt it as explicit technical debt, but not to waive it away with a "prove it" mentality. You can readily get into infinite Mexican stand-offs with that tack.

Clearly, we would prefer our design not to have this property. We're willing to accept it mostly because we have bigger fish to fry. I can construct example workloads where this will behave badly for sure. How much CPU will it burn? I can't say for sure. But I could get it to chase several hundred pointers per eviction pretty easily, which is definitely bad. It's likely a worse investment of our time to argue over the relevant standard of proof than it would be to address the issue.